### PR TITLE
WebAssembly cleanup

### PR DESCRIPTION
**Pull request by Cellule**
- Move Wasm related function from ScriptContext to WasmLibrary.
- Remove jsrt apis for WebAssembly.
- Only enable WebAssembly code in test binaries.
- Only set the `Wasm` global object with `-on:wasm` switch.
- Use array of Local for WasmSignature Params instead of growing array.
- Rename ModuleInfo.h to WasmModule.h
- Wasm::Local is no longer a struct, just a typedef to Wasm::WasmType.
- Remove namespace Binary.
- Merge WasmModule and ModuleInfo
- Merge WasmFunction and WasmFunctionInfo
- WasmBinaryReader is now allocated by WasmModule and their lifetime are bound together (how poetic)
- Merged WasmBinaryOpCodes.h and WasmKeywords.h Removed WasmBinOp in favor of WasmOp

This change is [Reviewable](https://reviewable.io/reviews/microsoft/chakracore/1324)

### CODE REVIEW

**Comment by Cellule** [comment]
Reviewed 35 of 35 files at r1.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[lib/Runtime/Base/CrossSite.cpp, line 312 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/1324#-KNK_yORiNPz6Y7s_LRV:-KNK_yORiNPz6Y7s_LRW:-1792565725) ([raw file](https://github.com/microsoft/chakracore/blob/101f77d21ed707d485542c907a879c977c5d0ef9/lib/Runtime/Base/CrossSite.cpp#L312)):*
> ```cpp
>                 funcInfo->GetFunctionBody()->GetIsAsmJsFunction())
>             {
> #ifdef ENABLE_WASM
> ```

That was a bug in the previous change. It would have failed to build with `ENABLE_WASM` not defined and `ASMJS_PLAT` defined

---

*[lib/Runtime/Library/JavascriptLibrary.cpp, line 1557 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/1324#-KNKaDl39UafcMjcsoh7:-KNKaDl39UafcMjcsoh8:-637086537) ([raw file](https://github.com/microsoft/chakracore/blob/101f77d21ed707d485542c907a879c977c5d0ef9/lib/Runtime/Library/JavascriptLibrary.cpp#L1557)):*
> ```cpp
>         nullEnumerator = RecyclerNew(this->recycler, NullEnumerator, scriptContext);
> #ifdef ENABLE_WASM
>         if (PHASE_ON1(WasmPhase))
> ```

This is what makes `-on:wasm` necessary to use WebAssembly now

---

*[lib/Runtime/Library/WasmLibrary.cpp, line 160 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/1324#-KNKbAJ0JIrP0z2yK_LN:-KNKbAJ0JIrP0z2yK_LO:1350152690) ([raw file](https://github.com/microsoft/chakracore/blob/101f77d21ed707d485542c907a879c977c5d0ef9/lib/Runtime/Library/WasmLibrary.cpp#L160)):*
> ```cpp
>         try
>         {
>             Wasm::WasmBytecodeGenerator::GenerateFunctionBytecode(scriptContext, readerInfo);
> ```

No longer pass the function body as parameter

---

*[lib/Runtime/Library/WasmLibrary.cpp, line 195 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/1324#-KNKbHwYz5CRsXyRC0Ok:-KNKbHwYz5CRsXyRC0Ol:718019792) ([raw file](https://github.com/microsoft/chakracore/blob/101f77d21ed707d485542c907a879c977c5d0ef9/lib/Runtime/Library/WasmLibrary.cpp#L195)):*
> ```cpp
>         for (uint i = 0; i < wasmModule->GetFunctionCount(); ++i)
>         {
>             AsmJsScriptFunction * funcObj = ctx->GetLibrary()->CreateAsmJsScriptFunction(wasmModule->GetFunctionInfo(i)->GetBody());
> ```

Changed `functionArray[i]->body` to `wasmModule->GetFunctionInfo(i)->GetBody()`

---

*[lib/Runtime/Library/WasmLibrary.cpp, line 496 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/1324#-KNKbVGixudc-g8JEu2F:-KNKbVGixudc-g8JEu2G:1206087291) ([raw file](https://github.com/microsoft/chakracore/blob/101f77d21ed707d485542c907a879c977c5d0ef9/lib/Runtime/Library/WasmLibrary.cpp#L496)):*
> ```cpp
> #endif
> 
> Js::JavascriptMethod Js::WasmLibrary::WasmDeferredParseEntryPoint(Js::AsmJsScriptFunction** funcPtr, int internalCall)
> ```

I had to put this function outside of the `#ifdef ENABLE_WASM` because I couldn't surround the callsite in the amd64 thunk (ENABLE_WASM is never defined in that file). I wonder how come this is not needed for asm.js functions called from the thunks

---

*[lib/WasmReader/Chakra.WasmReader.vcxproj, line 37 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/1324#-KNKcC1LCGyBoE_7Puts:-KNKcC1LCGyBoE_7Putt:-963726050) ([raw file](https://github.com/microsoft/chakracore/blob/101f77d21ed707d485542c907a879c977c5d0ef9/lib/WasmReader/Chakra.WasmReader.vcxproj#L37)):*
> $(MSBuildThisFileDirectory)

This is needed for razzle build, we just forgot to put it there in the first place

---

*[lib/WasmReader/WasmByteCodeGenerator.cpp, line 260 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/1324#-KNKdGhreDf9sBh_rJNx:-KNKdGhreDf9sBh_rJNy:-1894770883) ([raw file](https://github.com/microsoft/chakracore/blob/517e23d6567bbe33d8ae86a5b0f1d90a59ccc691/lib/WasmReader/WasmByteCodeGenerator.cpp#L260)):*
> ```cpp
>     info->SetArgsSizesArray(argSizeArray);
> 
>     if (m_module->memSize > 0)
> ```

I think this was a bug before that would always set UsesHeapBuffer to true

---

*[lib/WasmReader/WasmModule.cpp, line 269 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/1324#-KNKe068XoGPyLBrAyX7:-KNKe068XoGPyLBrAyX8:-1872453442) ([raw file](https://github.com/microsoft/chakracore/blob/101f77d21ed707d485542c907a879c977c5d0ef9/lib/WasmReader/WasmModule.cpp#L269)):*
> ```cpp
> void WasmModule::Finalize(bool isShutdown)
> {
>     m_alloc.Clear();
> ```

This was a bug, The recycler doesn't call the destructor on the object, therefor, the memory in the ArenaAllocator was leaked.

---

*[lib/WasmReader/WasmParseTree.h, line 16 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/1324#-KNKeChWvSdD2EoyYgx5:-KNKeChWvSdD2EoyYgx6:-1618480515) ([raw file](https://github.com/microsoft/chakracore/blob/101f77d21ed707d485542c907a879c977c5d0ef9/lib/WasmReader/WasmParseTree.h#L16)):*
> ```objc
>             // based on binary format encoding values
>             Void = 0,
>             I32 = 1,
> ```

I decided it wasn't worth using a `#define` + `#include` for 4 types, but if you think it's best to keep it that way I don't mind changing it back

---

*[lib/WasmReader/WasmParseTree.h, line 29 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/1324#-KNKeQ1jvyuUhLZOjRgb:-KNKeQ1jvyuUhLZOjRgc:553327902) ([raw file](https://github.com/microsoft/chakracore/blob/101f77d21ed707d485542c907a879c977c5d0ef9/lib/WasmReader/WasmParseTree.h#L29)):*
> ```objc
> #include "WasmBinaryOpcodes.h"
> 
>     enum WasmOp : byte
> ```

I realized that right now we only need up to `0xb9` opcodes and we could save some space by using only a byte

---

*[lib/WasmReader/WasmParseTree.h, line 117 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/1324#-KNKeXY6w_dzI2rU2i5B:-KNKeXY6w_dzI2rU2i5C:-365189246) ([raw file](https://github.com/microsoft/chakracore/blob/517e23d6567bbe33d8ae86a5b0f1d90a59ccc691/lib/WasmReader/WasmParseTree.h#L117)):*
> ```objc
> 
> }
> ```

This should have been removed along with sexpr parser

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/chakracore/1324)*

**Comment by Cellule** [comment]
@MikeHolman  @commonlisp Can you review this please ?

File: lib/Common/CommonDefines.h [hunk 1]
```diff
@@ -558,7 +558,8 @@
 #endif
 #endif
 
-#ifdef ASMJS_PLAT
+#if defined(ASMJS_PLAT) && defined(ENABLE_DEBUG_CONFIG_OPTIONS)
+// Enable WebAssembly only in debug and test build for the time begin
```
Comment (MikeHolman) [inline lib/Common/CommonDefines.h]: nit: spelling

File: lib/WasmReader/WasmBinaryOpCodes.h [hunk 2]
```diff
@@ -5,244 +5,255 @@
...
 #endif
 
 #ifndef WASM_MEM_OPCODE
-#define WASM_MEM_OPCODE(opname, opcode, token, sig) WASM_OPCODE(opname, opcode, token, sig)
+#define WASM_MEM_OPCODE(opname, opcode, sig, nyi) WASM_OPCODE(opname, opcode, sig, nyi)
 #endif
 
-#ifndef WASM_SIMPLE_OPCODE
-#define WASM_SIMPLE_OPCODE(opname, opcode, token, sig) WASM_OPCODE(opname, opcode, token, sig)
+#ifndef WASM_MEMREAD_OPCODE
```
Comment (MikeHolman) [inline lib/WasmReader/WasmBinaryOpCodes.h]: good! merging these was something i'd been wanting ever since there were 2 lists

**Comment by MikeHolman** [comment]
:shipit:

File: lib/Runtime/Library/WasmLibrary.cpp [hunk 3]
```diff
@@ -63,23 +68,20 @@ namespace Js
...
-                true, // isBinary
                 byteLength,
                 Js::Constants::GlobalCode,
                 ffi,
                 &start
             );
         END_LEAVE_SCRIPT_INTERNAL(scriptContext)
 
-        HRESULT hr = se.ei.scode;
+            HRESULT hr = se.ei.scode;
```
Comment (commonlisp) [inline lib/Runtime/Library/WasmLibrary.cpp]: Nit: indented too far?

File: lib/Runtime/Library/WasmLibrary.cpp [hunk 4]
```diff
@@ -98,5 +100,412 @@ namespace Js
...
+#elif defined(_M_X64)
+    // Do nothing: the implementation of WasmLibrary::WasmDeferredParseExternalThunk is declared (appropriately decorated) in
+    // Language\amd64\amd64_Thunks.asm.
+#endif
+
 }
-#endif
\ No newline at end of file
+
+#endif
```
Comment (commonlisp) [inline lib/Runtime/Library/WasmLibrary.cpp]: Would it be helpful to annotate with the closed block name since some of these #ifdefs are nested?

**Pull request merged**